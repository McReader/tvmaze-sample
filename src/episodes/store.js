import { createEntityAdapter, createAsyncThunk, createSlice } from "@reduxjs/toolkit";

const PREFIX = "episodes";

export const fetchShowEpisodeList = createAsyncThunk(`${PREFIX}/fetchShowEpisodeList`, async (showId, thunkAPI) => {
    const result = await thunkAPI.extra.showsService.getShowEpisodeList(showId);
    return result.data;
});

export const fetchEpisodeMainInformation = createAsyncThunk(`${PREFIX}/fetchEpisodeMainInformation`, async ({ showId, episodeId }, thunkAPI) => {
    const episode = episodesSelectors.selectById(thunkAPI.getState(), episodeId);

    if (episode) {
        return Promise.resolve(episode);
    }

    const result = await thunkAPI.extra.episodesService.getEpisodeMainInformation(episodeId);
    return result.data;
});

const episodesAdapter = createEntityAdapter();

export const initialState = {
    ...episodesAdapter.getInitialState(),
    byShow: {},
};

export const episodesSlice = createSlice({
    name: PREFIX,
    initialState: initialState,
    reducers: {
        // standard reducer logic, with auto-generated action types per reducer
    },
    extraReducers: {
        [fetchShowEpisodeList.fulfilled]: (state, { meta, payload }) => {
            episodesAdapter.addMany(state, payload);
            state.byShow[meta.arg] = payload.map(episode => episode.id);
        },
        [fetchEpisodeMainInformation.fulfilled]: (state, { meta, payload }) => {
            episodesAdapter.addOne(state, payload);
            state.byShow[meta.arg.showId] = [payload.id];
        },
    }
});

export const episodesSelectors = {
    ...episodesAdapter.getSelectors((state) => state.episodes),
    selectByShowId: (state) => {  },
};
