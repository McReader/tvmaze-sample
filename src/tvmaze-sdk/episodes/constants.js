export const EPISODES_RESOURCE_PATH = "/episodes";
