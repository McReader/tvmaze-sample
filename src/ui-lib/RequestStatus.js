import PropTypes from "prop-types";
import { CircularProgress } from "@material-ui/core";
import ErrorIcon from '@material-ui/icons/Error';

export const REQUEST_STATUS_IDLE = "idle";
export const REQUEST_STATUS_PENDING = "pending";
export const REQUEST_STATUS_FULFILLED = "fulfilled";
export const REQUEST_STATUS_REJECTED = "rejected";

/**
 * Renders progress bar when "requestStatus" is "idle" or "pending"
 * Renders caution symbol when "requestStatus" is "rejected"
 * Uses "render prop" pattern to render children when "request status" is "fulfilled"
 */
export function RequestStatus({ children, requestStatus = REQUEST_STATUS_IDLE }) {
    if (requestStatus === REQUEST_STATUS_IDLE || requestStatus === REQUEST_STATUS_PENDING) {
        return <CircularProgress  />;
    }

    if (requestStatus === REQUEST_STATUS_REJECTED) {
        return <ErrorIcon />;
    }

    return children();
}

export const requestStatusPropType = PropTypes.oneOf([
    REQUEST_STATUS_IDLE,
    REQUEST_STATUS_PENDING,
    REQUEST_STATUS_FULFILLED,
    REQUEST_STATUS_REJECTED
]);

RequestStatus.propTypes = {
    children: PropTypes.func.isRequired,
    requestStatus: requestStatusPropType,
};
