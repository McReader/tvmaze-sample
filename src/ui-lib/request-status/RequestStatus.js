import PropTypes from "prop-types"
import ErrorIcon from "@material-ui/icons/Error"

import { ProgressPlaceholder } from "../progress-placeholder/ProgressPlaceholder"
import { FillParentPlaceholder } from "../fill-parent-placeholder/FillParentPlaceholder"

export const REQUEST_STATUS_IDLE = "idle"
export const REQUEST_STATUS_PENDING = "pending"
export const REQUEST_STATUS_FULFILLED = "fulfilled"
export const REQUEST_STATUS_REJECTED = "rejected"

/**
 * Renders progress bar when "requestStatus" is "idle" or "pending"
 * Renders caution symbol when "requestStatus" is "rejected"
 * Uses "render prop" pattern to render children when "request status" is "fulfilled"
 */
export function RequestStatus({
	children,
	requestStatus = REQUEST_STATUS_IDLE,
}) {
	if (
		requestStatus === REQUEST_STATUS_IDLE ||
		requestStatus === REQUEST_STATUS_PENDING
	) {
		return <ProgressPlaceholder />
	}

	if (requestStatus === REQUEST_STATUS_REJECTED) {
		return (
			<FillParentPlaceholder>
				<ErrorIcon />
			</FillParentPlaceholder>
		)
	}

	return children()
}

export const requestStatusPropType = PropTypes.oneOf([
	REQUEST_STATUS_IDLE,
	REQUEST_STATUS_PENDING,
	REQUEST_STATUS_FULFILLED,
	REQUEST_STATUS_REJECTED,
])

RequestStatus.propTypes = {
	children: PropTypes.func.isRequired,
	requestStatus: requestStatusPropType,
}
