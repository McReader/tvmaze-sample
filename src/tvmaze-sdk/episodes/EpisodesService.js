import { EPISODES_RESOURCE_PATH } from "./constants";

export class EpisodesService {
    constructor(httpClient) {
        this.httpClient = httpClient;
    }

    /**
     * Retrieve all primary information for a given episode.
     * @param {number} episodeId 
     * @return {Promise<Episode>}
     */
    getEpisodeMainInformation(episodeId) {
        return this.httpClient.get(`${EPISODES_RESOURCE_PATH}/${episodeId}`);
    }
}
