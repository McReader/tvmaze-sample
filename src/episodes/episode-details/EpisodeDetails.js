import { Details, detailsPropTypes } from "../../ui-lib/details/Details"
import { requestStatusPropType } from "../../ui-lib/request-status/RequestStatus"

export function EpisodeDetails({ episode, requestStatus }) {
	return <Details details={episode} requestStatus={requestStatus} />
}

EpisodeDetails.propTypes = {
	episode: detailsPropTypes,
	requestStatus: requestStatusPropType,
}
