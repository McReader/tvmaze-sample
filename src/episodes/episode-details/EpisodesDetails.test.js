import renderer from "react-test-renderer";

import { EpisodeDetails } from "./EpisodeDetails";

import episodes from "../../tvmaze-sdk/episodes/__mocks__/episodes.json";

import { REQUEST_STATUS_FULFILLED } from "../../ui-lib/request-status/RequestStatus";

let tree;
let props;

beforeEach(() => {
    tree = renderer
        .create(<EpisodeDetails {...props} />)
        .toJSON();
});

describe(`when request status is "${REQUEST_STATUS_FULFILLED}" and there is episode`, () => {
    beforeAll(() => {
        props = {
            requestStatus: REQUEST_STATUS_FULFILLED,
            episode: episodes[0],
        };
    });

    test("should match snapshot", () => {
        expect(tree).toMatchSnapshot();
    });
});

