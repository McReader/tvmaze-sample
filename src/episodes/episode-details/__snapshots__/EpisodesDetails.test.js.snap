// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when request status is "fulfilled" and there is episode should match snapshot 1`] = `
<div
  className="Details"
>
  <h1>
    Pilot
  </h1>
  <div
    className="DetailsSummaryContainer"
  >
    <picture
      className="DetailsCoverImage"
    >
      <source
        media="(max-width:600px)"
        srcSet="http://static.tvmaze.com/uploads/images/original_untouched/1/4388.jpg"
      />
      <source
        media="(man-width:900px)"
        srcSet="http://static.tvmaze.com/uploads/images/medium_landscape/1/4388.jpg"
      />
      <img
        alt="Cover"
        className="CoverImageImg"
        src="http://static.tvmaze.com/uploads/images/original_untouched/1/4388.jpg"
      />
    </picture>
    <summary
      className="DetailsSummary"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>When the residents of Chester's Mill find themselves trapped under a massive transparent dome with no way out, they struggle to survive as resources rapidly dwindle and panic quickly escalates.</p>",
        }
      }
    />
  </div>
</div>
`;
